typegen/tuple.go

```diff
@@ -27,9 +27,34 @@ func (tg *TypeGenerator) GenTuple(tup *types.Si1TypeDefTuple, mt *types.Portable
 		return g, nil
 	}
 
-	// We just name tuples based off their metadata id, as that is guaranteed to be unique, and
-	// could help with debugging
-	tn := utils.AsName("Tuple", fmt.Sprint(mt.ID.Int64()))
+	var tn string
+	if len(*tup) == 2 {
+		// Name tuples of only 2 elements based on their interior elements and order
+		// TupleOf{Type1Name}{Type2Name}
+		g1, err := tg.GetType((*tup)[0].Int64())
+		if err != nil {
+			return nil, err
+		}
+		g2, err := tg.GetType((*tup)[1].Int64())
+		if err != nil {
+			return nil, err
+		}
+		// Although this name is almost certainly unique, it may still not be totally unique, so we
+		// may need to add an integer postfix
+		tn = utils.AsName("TupleOf", g1.DisplayName(), g2.DisplayName())
+
+		if tg.nameCount[tn] == 0 {
+			tg.nameCount[tn] = 1
+		} else {
+			tg.nameCount[tn] += 1
+			tn = utils.AsName(tn, fmt.Sprint(tg.nameCount[tn]-1))
+		}
+	} else {
+		// We name longer tuples based off their metadata id, as that is guaranteed to be unique, and
+		// could help with debugging
+		tn = utils.AsName("Tuple", fmt.Sprint(mt.ID.Int64()))
```

Comment:
Should we just also do this for `len(*tup) == 2 && tg.nameCount[tn] != 0`? Probably better than using `tg.nameCount[tn] - 1` as the number.

Reply:
Well if we have `len(*tup) == 2 && tg.nameCount[tn] == n`, we'll end up with a name of:

`TupleOf{Type1}{Type2}{n-1}`

In my opinion that's still preferable because it does describe the contained types. That said, I can't think of a set of types in which this would happen unless there's just a naming collision on a named type. In that case (and in the case of `Tuple{xxx}`), you would need to disambiguate via the actual code name of the contained type.

Thoughts?

Reply:
Yeah that's fair!